### PR TITLE
make affine parallel and yield MemRef normalizable

### PR DESCRIPTION
**Pull request by adstraw**

### CODE REVIEW

File: mlir/lib/Transforms/NormalizeMemRefs.cpp [hunk 1]
```diff
@@ -517,6 +513,10 @@ Operation *NormalizeMemRefs::createOpResultsNormalized(FuncOp funcOp,
   // affine map, `oldOp` is returned without modification.
   if (resultTypeNormalized) {
     OpBuilder bb(oldOp);
+    for (auto &oldRegion : oldOp->getRegions()) {
+      auto newRegion = result.addRegion();
+      newRegion->getBlocks().splice(newRegion->begin(), oldRegion.getBlocks());
```
Comment (flaub) [inline mlir/lib/Transforms/NormalizeMemRefs.cpp]: Maybe 
```
Region* newRegion = result.addRegion();
newRegion->takeBody(oldRegion);
```

Note that the MLIR community would rather us spell out the `auto` unless the RHS already has the type (like it's a dyn_cast or something).

**Review by flaub — APPROVED** [review]
LGTM, minor comment about using `takeBody`

File: mlir/test/Transforms/normalize-memrefs.mlir [hunk 2]
```diff
@@ -319,3 +319,16 @@ func @use_value_of_external(%A: memref<16xf64, #tile>, %B: f64) -> (memref<8xf64
 }
 // CHECK: %[[res:[0-9]+]] = call @external_func_B(%[[A]], %[[B]]) : (memref<4x4xf64>, f64) -> memref<2x4xf64>
 // CHECK: return %{{.*}} : memref<2x4xf64>
+
+// CHECK-LABEL: func @affine_parallel_norm
+func @affine_parallel_norm() ->  memref<8xf32, #tile> {
+  %c = constant 23.0 : f32
+  // CHECK: %{{.*}} = affine.parallel (%{{.*}}) = (0) to (8) reduce ("assign") -> (memref<2x4xf32>)
+  %1 = affine.parallel (%i) = (0) to (8) reduce ("assign") ->  memref<8xf32, #tile> {
+    %a = alloc() : memref<8xf32, #tile>
```
Comment (flaub) [inline mlir/test/Transforms/normalize-memrefs.mlir:328]: While this code is fine, it's a slight bit odd to be allocating a buffer and storing into it then yielding it to the outside scope. For clarity, hoist the alloc to the function-level scope.

Comment (adstraw) [reply mlir/test/Transforms/normalize-memrefs.mlir:328]: Done.

File: mlir/lib/Transforms/NormalizeMemRefs.cpp (hunk 1 above)
Comment (adstraw) [reply mlir/lib/Transforms/NormalizeMemRefs.cpp]: Done.

**Pull request merged**